pkg/pool-weighted/contracts/smart/ManagedPool.sol

```diff
@@ -444,6 +445,268 @@ contract ManagedPool is BaseWeightedPool, ProtocolFeeCache, ReentrancyGuard {
+        // Let Sb, Wnb = total supply and normalized weight before the add; and
+        //     Sa, Wna = total supply and normalized weight after the add
+        // Then: (Sa * Wna) / B = (Sb * Wnb) / B
+        //       Sa * Wna = Sb * Wnb
+        //       Sa = Sb * (Wnb / Wna)
+        // Since the normalized weights (Wn) = denormalized weights (Wd) / weightSum (WS), we have:
+        // Sa = Sb * (Wd / WSb)  - denormalized weights don't change, so there is just one Wd
+        //            ---------
+        //           (Wd / WSa)
+        // Sa = Sb * WSa / WSb = totalSupply after the add
+        //
+        // Then the bptAmountOut is the delta in the total supply:
+        // bptAmountOut = Sa - Sb
+        //              = Sb * WSa / WSb - Sb
+        //              = Sb * (WSa / WSb - 1)
+        //
+        // In our example, Sa is the totalSupply on initialization of the pool. If the balances are 400 and 0.5:
+        // Sb = (400^0.8) * (0.5^0.2) * 2 = 210.1222
+        // Sa = 210.1222 * (2.5 / 1.0) = 525.3056
+        // bptAmountOut = 210.1222 * (2.5 / 1.0 - 1) = 315.1833
+        // The added token should represent 60% of the total supply, after the add, and in fact:
+        // 315.1833 / 525.3056 = ~ 0.6
+        //
+        uint256 weightSumRatio = weightSumAfterAdd.divDown(weightSumBeforeAdd);
+        bptAmountOut = totalSupply().mulDown(weightSumRatio.sub(FixedPoint.ONE));
+
+        // Validate that the actual BPT price
+        uint256 actualBptPrice = totalSupply().mulDown(weightSumRatio).mulDown(normalizedWeight).divUp(
+            _upscale(tokenAmountIn, _computeScalingFactor(token))
+        );
+
+        _require(actualBptPrice >= minBptPrice, Errors.MIN_BPT_PRICE_ADD_TOKEN);
+    }
+
+    function _registerNewToken(
+        IERC20 token,
+        uint256 denormalizedWeight,
+        uint256 tokenAmountIn,
+        address assetManager
+    )
+        private
+        returns (
+            IERC20[] memory,
+            uint256,
+            uint256[] memory
+        )
+    {
+        address[] memory assetManagers = new address[](1);
+        assetManagers[0] = assetManager;
+        IERC20[] memory tokensToAdd = new IERC20[](1);
+        tokensToAdd[0] = token;
+
+        getVault().registerTokens(getPoolId(), tokensToAdd, assetManagers);
+
+        // Tokens array is now different
+        (IERC20[] memory tokens, , ) = getVault().getPoolTokens(getPoolId());
+        uint256[] memory maxAmountsIn = new uint256[](tokens.length);
+        uint256 tokenIndex;
+
+        // It will put it at the end, so iterate backwards to find it faster
+        for (uint256 i = tokens.length - 1; i >= 0; i--) {
+            if (tokens[i] == token) {
+                // This is the new one we're joining with
+                maxAmountsIn[i] = tokenAmountIn;
+                tokenIndex = i;
+                break;
+            }
+        }
```

Comment:
This is unnecessary, since as noted the token will be at the end of the array.

However, this made me notice a bug. We derive BaseWeightedPool, which selects the pool type based on the number of tokens: minimal swap info if >2, and two token if == 0. This is wrong: a two token pool cannot add or remove tokens, we should use minimal swap info always.

Reply:
Good catch.

Reply:
I've added a `mutableTokens` arg to `BaseWeightedPool` to allow us to override this behaviour in [3d63cb3](https://github.com/balancer-labs/balancer-v2-monorepo/pull/1131/commits/3d63cb32876c6f55192124c837223ee7482adb33)